### DISCUSSION TITLE
### How to restore main branch to previous state after wrong merge

### DISCUSSION THREAD

**Discussion by haroshi-hub** [community/community#144684] [Codespaces] [answered]
### Select Topic Area

Question

#### What I'm trying to do
I need to revert my main branch to its previous state after accidentally merging changes from branch 'A' that shouldn't have been merged.

#### Current situation
1. Created a new branch 'A'
2. Made and pushed new commits to branch 'A'
3. Merged branch 'A' into main branch
4. Realized the merge was incorrect and need to undo it

#### What I need help with
- How to safely restore the main branch to its state before the merge
- Best practice to handle this situation without losing commit history

#### Environment
- Using GitHub Web Interface/Git CLI (please specify which you're using)

#### Additional Information
- Are there any other branches or concurrent changes I should be aware of?
- Have any other team members pulled the updated main branch?

Please let me know the safest way to handle this situation.

**Comment by wlb-tycoon** [comment, marked as answer]
There are several ways to restore your main branch to its previous state. I'll provide both simple and advanced solutions:

### 1. Using `git reset` (If changes haven't been pushed to remote)
```
git checkout main
git reset --hard HEAD~1   # Goes back 1 commit (the merge commit)
```

### 2. Using git revert (Safer option, especially if changes are already pushed)
```
git checkout main
git revert -m 1 <merge-commit-hash>
```

### 3. Step-by-step solution (Recommended approach)
- First, find the merge commit:
```
git log --merges
```
- Choose your preferred method:
```
For local-only changes: Use reset method (#1)
For pushed changes: Use revert method (#2)
```
- After fixing, push the changes:
```
git push origin main --force  # For reset method
# OR
git push origin main         # For revert method
```
